rotkehlchen/tasks/historical_balances.py

```diff
@@ -0,0 +1,242 @@
+                protocol=counterparty,
+                asset=asset,
+            )
+
+        return cls(
+            location=location,
+            location_label=event.location_label,
+            protocol=None,
+            asset=asset,
+        )
+
+
+def _load_bucket_balances_before_ts(
+        database: 'DBHandler',
+        from_ts: Timestamp,
+) -> dict[Bucket, FVal]:
+    """Load the latest balance per bucket from event_metrics for events before from_ts.
+
+    For each bucket (location, location_label, protocol, asset), we need the balance
+    from the last event before from_ts. Query is ordered by timestamp DESC so first
+    occurrence per bucket is the latest balance.
+    """
+    bucket_balances: dict[Bucket, FVal] = {}
+    with database.conn.read_ctx() as cursor:
+        cursor.execute(
+            """
+            SELECT he.location, he.location_label, em.protocol, he.asset, em.metric_value
+            FROM event_metrics em
+            INNER JOIN history_events he ON em.event_identifier = he.identifier
+            WHERE em.metric_key = ? AND he.timestamp < ?
+            ORDER BY he.timestamp DESC, he.sequence_index DESC
+            """,
+            (EventMetricKey.BALANCE, from_ts),
+        )
+        for row in cursor:
+            if (bucket := Bucket.from_db(row[:4])) not in bucket_balances:
+                bucket_balances[bucket] = FVal(row[4])
+
+    log.debug(f'Loaded {len(bucket_balances)} bucket balances before ts={from_ts}')
+    return bucket_balances
+
+
+def process_historical_balances(
+        database: 'DBHandler',
+        msg_aggregator: 'MessagesAggregator',
+        from_ts: Timestamp | None = None,
+) -> None:
+    """Process events and compute balance metrics. Stops on negative balance."""
+    log.debug(f'Starting historical balance processing from_ts={from_ts}')
+    bucket_balances = {}
+    if from_ts is not None:
+        bucket_balances = _load_bucket_balances_before_ts(database, from_ts)
+
+    with database.conn.read_ctx() as cursor:
+        filter_query = HistoryEventFilterQuery.make(
+            from_ts=from_ts,
+            order_by_rules=[('timestamp', True), ('sequence_index', True)],
+            exclude_ignored_assets=True,
+        )
+        events = DBHistoryEvents(database).get_history_events_internal(
+            cursor=cursor,
+            filter_query=filter_query,
+        )
+
+    if (total_events := len(events)) == 0:
+        log.debug('No events to process for historical balances')
+        _finalize_processing(database)
+        return
+
+    metrics_batch: list[tuple[int | None, str | None, str, str]] = []
+    first_batch_written, negative_balance_buckets = False, set[Bucket]()
+    send_ws_every = msg_aggregator.how_many_events_per_ws(total_events)
+    for idx, event in enumerate(events):
+        if (bucket := Bucket.from_event(event)) in negative_balance_buckets:
+            continue
+
+        current_balance_in_bucket = bucket_balances.get(bucket, ZERO)
```

Comment:
why do you need the set. isnt it better to check if current_balance_in_bucket is negative?